pandas/io/excel/_base.py

```diff
@@ -771,15 +772,14 @@ def _value_with_fmt(self, val):
 
         return val, fmt
 
-    @classmethod
-    def check_extension(cls, ext):
+    def check_extension(self, ext):
```

Comment:
@WillAyd 

AFAICT the implementation works but mypy is reporting `pandas\io\excel\_base.py:781:54: error: "Callable[[ExcelWriter], Tuple[str, ...]]" has no attribute "__iter__" (not iterable)  [attr-defined]` (now L780)

although `supported_extensions` is a class variable, and it is defined in the abstract base class as a property purely to ensure that a subclass defines it. mypy thinks that it may also be a property in a subclass and this would be invalid code.

asked on gitter https://gitter.im/python/typing?at=5db30d223d669b28a0d62c6f but response was that mypy is correct.

`check_extension` does not need to be a class method and mypy does not report errors with a regular method.

any reason not to change from classmethod?

Reply:
Well the downside is that this is technically exposed as part of the API, so moving from a class method to an instance method would be a breaking change.

For the subclasses does defining `supported_extensions` as a property not fix it? Glanced at the mypy convo that was what I picked up but might have misinterpreted

Reply:
> For the subclasses does defining `supported_extensions` as a property not fix it?

IIUC that is why mypy is reporting an error, since defining `supported_extensions` as a property in the subclasses leads to a runtime error...

The other option is not to define `supported_extensions` as an abstract property.

```
__________________________________________________________________________________________ test_excel_multiindex_index[.xls] __________________________________________________________________________________________ 

ext = '.xls'

    def test_excel_multiindex_index(ext):
        # MultiIndex as index works so assert no error #9794
        cols = MultiIndex.from_tuples(
            [("site", ""), ("2014", "height"), ("2014", "weight")]
        )
        df = DataFrame(np.random.randn(3, 10), index=cols)
        with ensure_clean(ext) as path:
>           df.to_excel(path, index=False)

pandas\tests\io\excel\test_xlwt.py:42:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas\core\generic.py:2296: in to_excel
    engine=engine,
pandas\io\formats\excel.py:735: in write
    writer = ExcelWriter(_stringify_path(writer), engine=engine)  # type: ignore
pandas\io\excel\_xlwt.py:25: in __init__
    super().__init__(path, mode=mode, **engine_kwargs)
pandas\io\excel\_base.py:714: in __init__
    self.check_extension(ext)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.io.excel._xlwt._XlwtWriter'>, ext = 'xls'

    @classmethod
    def check_extension(cls, ext):
        """checks that path's extension against the Writer's supported
        extensions.  If it isn't supported, raises UnsupportedFiletypeError."""
        if ext.startswith("."):
            ext = ext[1:]
>       if not any(ext in extension for extension in cls.supported_extensions):
E       TypeError: 'property' object is not iterable

pandas\io\excel\_base.py:781: TypeError
```

Reply:
This code is pretty strange. Does defining `a` as an abstract `classmethod` instead of a `property` work? What we have in the base class doesn't work unless you overwrite the property with a class variable, so yea I think makes sense for mypy to flag this

Reply:
> What we have in the base class doesn't work unless you overwrite the property with a class variable

IIUC what we have in the base class is the construct to ensure that subclasses define `supported_extensions`, so i'm not sure what you mean when you say what is in the base class doesn't work since it is an abstract class that cannot be instantiated.

so yes "What we have in the base class doesn't work unless you overwrite the property with a class variable" but isn't that the reason for using

```
    @property
    @abc.abstractmethod
    def supported_extensions(self):
        """Extensions that writer engine supports."""
        pass
```
instead of just 
```
supported_extensions = None
```

Reply:
`property` I believe works for only for instances; we are just mixing instance / class methods in a strange way and mypy is letting us know that. If you ignore the abstract thing for a minute, it certainly would be an error to call a class method that depends on an instance method

Sure you could just remove the `abstract` requirement altogether for `supported_extensions` and make it a class variable, but that would lose the abstract nature of it. You can alternately just convert the property to a class method and that keeps things consistent while passing validation

```python
from abc import ABCMeta, abstractmethod
from typing import Tuple


class A(metaclass=ABCMeta):
    @classmethod
    @abstractmethod
    def a(self) -> Tuple[int, ...]:
        pass

    @classmethod
    def b(cls) -> None:
        for x in cls.a():
            pass


class B(A):

    @classmethod
    def a(self):
        return (1, 2, 3)
```

Reply:
> You can alternately just convert the property to a class method and that keeps things consistent while passing validation

Thanks @WillAyd for looking into this.

i _think_ that this suggestion would also be considered a breaking change.

```
___________________________________________________________________________________ TestExcelWriterEngineTests.test_register_writer ___________________________________________________________________________________ 

self = <pandas.tests.io.excel.test_writers.TestExcelWriterEngineTests object at 0x00000202B49F78C8>

    def test_register_writer(self):
        # some awkward mocking to test out dispatch and such actually works
        called_save = []
        called_write_cells = []
    
        class DummyClass(ExcelWriter):
            called_save = False
            called_write_cells = False
            supported_extensions = ["xlsx", "xls"]
            engine = "dummy"
    
            def save(self):
                called_save.append(True)
    
            def write_cells(self, *args, **kwargs):
                called_write_cells.append(True)
    
        def check_called(func):
            func()
            assert len(called_save) >= 1
            assert len(called_write_cells) >= 1
            del called_save[:]
            del called_write_cells[:]
    
        with pd.option_context("io.excel.xlsx.writer", "dummy"):
            register_writer(DummyClass)
>           writer = ExcelWriter("something.xlsx")

pandas\tests\io\excel\test_writers.py:1257:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas\io\excel\_base.py:705: in __init__
    self.check_extension(ext)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.tests.io.excel.test_writers.TestExcelWriterEngineTests.test_register_writer.<locals>.DummyClass'>, ext = 'xlsx'

    @classmethod
    def check_extension(cls, ext: str) -> bool:
        """checks that path's extension against the Writer's supported
        extensions.  If it isn't supported, raises UnsupportedFiletypeError."""
        if ext.startswith("."):
            ext = ext[1:]
>       if not any(ext in extension for extension in cls.supported_extensions()):
E       TypeError: 'list' object is not callable

pandas\io\excel\_base.py:771: TypeError
```

Reply:
> Well the downside is that this is technically exposed as part of the API, so moving from a class method to an instance method would be a breaking change.

"None of the methods and properties are considered public." see https://dev.pandas.io/docs/reference/api/pandas.ExcelWriter.html?highlight=excelwriter#pandas.ExcelWriter

Reply:
> AFAICT the implementation works but mypy is reporting `pandas\io\excel\_base.py:781:54: error: "Callable[[ExcelWriter], Tuple[str, ...]]" has no attribute "__iter__" (not iterable) [attr-defined]` (now L780)
> 
> although `supported_extensions` is a class variable, and it is defined in the abstract base class as a property purely to ensure that a subclass defines it. mypy thinks that it may also be a property in a subclass and this would be invalid code.
> 
> asked on gitter https://gitter.im/python/typing?at=5db30d223d669b28a0d62c6f but response was that mypy is correct.

xref https://github.com/python/mypy/issues/7760

Reply:
I dont know how either mypy or abc.abstractmethod would handle this, but it isn't _that_ difficult to implement a class-property (i previously looked into this for Timestamp/Timedelta min/max)